File: todolist/src/types/countdown.ts
/** 预设分类定义 */
export interface CategoryPreset {
  label: string;
  icon: string;
  color: string;
}

/** Rust 返回的原始事件数据 */
export interface CountdownEvent {
  id: string;
  title: string;
  target_date: string;
  category: string;
  repeat_yearly: boolean;
  show_on_dashboard: boolean;
  color?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

/** 创建参数 */
export interface CreateCountdownParams {
  title: string;
  target_date: string;
  category?: string;
  repeat_yearly?: boolean;
  show_on_dashboard?: boolean;
  color?: string;
  notes?: string;
}

/** 更新参数 */
export interface UpdateCountdownParams {
  title?: string;
  target_date?: string;
  category?: string;
  repeat_yearly?: boolean;
  show_on_dashboard?: boolean;
  color?: string;
  notes?: string;
}

/** 前端增强类型 — 含计算字段 */
export interface CountdownEventDisplay extends CountdownEvent {
  daysRemaining: number;
  displayText: string;      // "还剩 128 天" | "已过 89 天" | "今天"
  displayType: 'remaining' | 'passed' | 'today';
  categoryLabel: string;
  categoryIcon: string;
  categoryColor: string;
}

File: todolist/src-tauri/src/models/countdown.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CountdownEvent {
    pub id: String,
    pub title: String,
    pub target_date: String,
    pub category: String,
    pub repeat_yearly: bool,
    pub show_on_dashboard: bool,
    pub color: Option<String>,
    pub notes: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateCountdownParams {
    pub title: String,
    pub target_date: String,
    #[serde(default = "default_category")]
    pub category: String,
    #[serde(default)]
    pub repeat_yearly: bool,
    #[serde(default = "default_show")]
    pub show_on_dashboard: bool,
    pub color: Option<String>,
    #[serde(default)]
    pub notes: Option<String>,
}

fn default_category() -> String { "其他".into() }
fn default_show() -> bool { true }

#[derive(Debug, Deserialize)]
pub struct UpdateCountdownParams {
    pub title: Option<String>,
    pub target_date: Option<String>,
    pub category: Option<String>,
    pub repeat_yearly: Option<bool>,
    pub show_on_dashboard: Option<bool>,
    pub color: Option<String>,
    pub notes: Option<String>,
}
